perf_benchmark/benchmark_main.cc

```diff
@@ -0,0 +1,435 @@
+    // For a small benchmark script, this is okay.
+    auto* service = new google::api::Service();
+    GOOGLE_CHECK_OK(
+        LoadService(std::string(kServiceConfigTextProtoFile), service));
+
+    // Create a TypeHelper based on the service config.
+    // Construct object on the heap for the same reason as the Service config.
+    auto* type_helper = new TypeHelper(service->types(), service->enums());
+    return type_helper;
+  }();
+  return *kTypeHelper;
+}
+} // namespace
+
+// Helper method to run Json Translation benchmark.
+//
+// error.
+// state - ::benchmark::State& variable used for collecting metrics.
+// msg_type - Protobuf message name for translation.
+// json_msg - Complete input json message.
+// streaming - Flag for streaming testing. When true, a stream of `stream_size`
+//             number of `json_msg` will be fed into translation.
+// stream_size - Number of streaming messages.
+// chunk_per_msg - Number of data chunks per message.
+absl::Status BenchmarkJsonTranslation(::benchmark::State& state,
+                                      absl::string_view msg_type,
+                                      absl::string_view json_msg,
+                                      bool streaming,
+                                      uint64_t stream_size,
+                                      uint64_t chunk_per_msg) {
+  // Retrieve global type helper
+  const TypeHelper& type_helper = GetBenchmarkTypeHelper();
+
+  // Get message type
+  const google::protobuf::Type* type = type_helper.Info()->GetTypeByTypeUrl(
+      absl::StrFormat("type.googleapis.com/%s", msg_type));
+  if (nullptr == type) {
+    return absl::InvalidArgumentError(absl::StrCat(
+        "Could not resolve the message type ",
+        msg_type));
+  }
+
+  RequestInfo request_info;
+  // body field path used in this benchmark are all "*"
+  request_info.body_field_path = "*";
+  request_info.variable_bindings = std::vector<RequestWeaver::BindingInfo>();
+  request_info.message_type = type;
+
+  // Wrap json_msg inside BenchmarkZeroCopyInputStream.
+  std::unique_ptr<BenchmarkZeroCopyInputStream> is;
+  if (streaming) {
+    std::string streaming_msg = GetStreamedJson(json_msg, stream_size);
+    is = absl::make_unique<BenchmarkZeroCopyInputStream>(streaming_msg,
+                                                         chunk_per_msg);
+  } else {
+    is = absl::make_unique<BenchmarkZeroCopyInputStream>(std::string(json_msg),
+                                                         chunk_per_msg);
+  }
+
+  // Benchmark the transcoding process
+  std::string message;
+  for (auto s: state) {
+    JsonRequestTranslator translator
+        (type_helper.Resolver(), is.get(), request_info, streaming, false);
+    MessageStream& out = translator.Output();
+
+    if (!out.Status().ok()) {
+      return absl::InternalError(out.Status().ToString());
+    }
+
+    while (out.NextMessage(&message)) {
```

Comment:
Could you try to use `script/check-style` to check the  coding style?   Even it is still using clang-3.8.0 version, but it is better than nothing?

If you want, you can use separate pr to upgrade this script to use latest clang version for code formatting.

Reply:
Thanks done! I can look into upgrading clang-format to a more recent version. Currently, clang-format doesn't provide x86_64 linux pre-build binary. We need a little bit more work than just changing the version number.